### ISSUE TITLE
Introducing Guava library to the project

### ISSUE THREAD

**Issue by maskaravivek** [commons-app/apps-android-commons#662]
I was hoping to introduce the guava library to the project. A [ lot of android apps](http://www.appbrain.com/stats/libraries/details/guava/google-guava) use it because of the many features that it provides. Some relevant links. 
1. [Advantages of guava](http://stackoverflow.com/questions/3759440/the-guava-library-for-java-what-are-its-most-useful-and-or-hidden-features?rq=1)
2. [Is guava useful for android apps?](http://stackoverflow.com/questions/3759440/the-guava-library-for-java-what-are-its-most-useful-and-or-hidden-features?rq=1)
3. [Comparison of guava with other library](http://stackoverflow.com/questions/4542550/what-are-the-big-improvements-between-guava-and-apache-equivalent-libraries)

**Comment by maskaravivek** [comment]
@nicolas-raoul @misaochan @whym It would be good to have your views on it. :)

**Comment by whym** [comment]
What is the current status of this? com.google.guava:guava has been adeed to the dependencies, but I don't see it is used anywhere in the source code.

**Comment by maskaravivek** [comment]
@whym I used it in #691. Will add a wiki soon so that we can start using it instead of async tasks. That would help in handling concurrency in an elegant way. :)

**Comment by whym** [comment]
That would be great. I missed that because I didn't notice the namespace was com.google.common not something like com.google.guava.

**Comment by psh** [comment]
![Method Count Graph](https://hawke.io/apps-android-commons-methods-count.png)

Guava is a huge library and Google's guidance is to always use ProGuard if you're going to pull it in.  The current master branch is right at 60k methods (of the 64k method limit).  The green quadrant in the top right of the graph is 14.9k methods pulled in from Guava.  At present the Guava usage is just used for `Futures`, `SettableFuture`, `ListenableFuture`, and `FutureCallback`.

Another contribution to the large method code is the plethora of networking libraries - Volley, Legacy Apache and a transitive dependency on OkHttp.

The main use is to do network calls in the background, something that OkHttp does extremely well especially if coupled with RxJava.  I believe that a switch to using Rx would also address the downside of #749 (fix to category threading) using its 'debounce' operator.

Would anyone be opposed to a spike to
* Unify the networking
* Switch from Guava to Rx (reporting on method count reduction) to simplify threading 
* Debounce the category search so it doesn't "create a bazillion executors" (to quote @misaochan )

**Comment by misaochan** [comment]
Hi @psh ,

Thanks to the insight! I would not be opposed to switching to Rx, personally. I am not terribly familiar with either Rx or Guava, but if Rx can fulfill Guava's current role while also making it simpler to fix #749 , I would be all for it.

@maskaravivek what do you think?

**Comment by maskaravivek** [comment]
Thanks @psh for the comments. :)

I have no experience with Rx so will have to evaluate it a bit. Yes if it helps in reducing the method count then we can surely consider it.

**Comment by neslihanturan** [comment]
Hi @psh 
Thanks for the easy fix advise for #667 . Did you done anything related so far or can you please give me some hints about how can I solve it by using RxJava (I have no experience with it yet) . It can be very nice if we can solve the issue in a week or sooner.

**Comment by psh** [comment]
@neslihanturan The PR I submitted (#782 - For discussion: get upload count using Kotlin & Rx) has the effect of removing Guava and swapping it for Rx.  There's no hard requirement to include Kotlin: plain old Java 7 syntax works just fine, its just a little more verbose.  Another option is to adopt Java 8 syntax (simple change in the `build.gradle`) if we want nicer syntax.

If it helps with a decision then we can untangle the various decisions - 

1. Decide if we want to swap Guava for RxJava
2. Fix the category issue (either the [debounce](http://reactivex.io/documentation/operators/debounce.html) operator, or something else)
3. Decide if we want to introduce Kotlin at a later date and tidy up the code we wrote for step 1

My person preference is a "yes" to Kotlin and a "yes" to using Rx for whatever that's worth!

Sorry if I conflated too many things and made the decisions harder than they needed to be.
